### PR TITLE
feat(template): externalize cloning of target resource

### LINKED ISSUE #202: Template based clone operations

### PR DESCRIPTION
**Pull request by aslakknutsen**
Use Go Templates to generate JSON Patch(rfc6902) objects to apply
to the target resource during mutation.

fixes #202

- [x] Test template.JSON operations
- [x] Test template.Engine

### CODE REVIEW

File: pkg/k8s/deployment.go [hunk 1]
```diff
@@ -88,42 +92,31 @@ func DeploymentRevertor(ctx model.SessionContext, ref *model.Ref) error { //noli
...
-				FieldPath:  "metadata.namespace",
-			},
-		},
+	org, err := json.Marshal(deployment)
+	if err != nil {
+		return nil, err
+	}
+
+	e := template.NewDefaultEngine()
+	modified, err := e.Run("telepresence", org, version, map[string]string{
```
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply pkg/k8s/deployment.go]: The thinking is that the sessions object will be changed to :

```
spec:
  refs: [
    {name: deptarget, strategy: telepresence}
  ] 
```

As it is now that argument could be removed. But in a PR soon(tm) the client will control it to tell us 'what do you want to do?' tp, squash, ...

File: pkg/template/template_test.go [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+var _ = Describe("Operations for template system", func() {
+
+	Context("json object", func() {
+		var (
+			err error
+			tj  template.JSON
+		)
+
+		BeforeEach(func() {
+			tj, err = template.NewJSON([]byte(org))
```
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply pkg/template/template_test.go]: https://github.com/Maistra/istio-workspace/pull/206/commits/97b2db5bb48ef7108768564dfa5e9b7188320aeb

File: pkg/template/template_test.go [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+		})
+		Context("slice values", func() {
+			It("should have value in slice", func() {
+				v := tj.Has("/spec/template/spec/containers/0")
+				Expect(v).To(BeTrue())
+			})
+			It("should get value in slice", func() {
+				v, err := tj.Value("/spec/template/spec/containers/0")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(v).ToNot(BeNil())
```
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply pkg/template/template_test.go]: https://github.com/Maistra/istio-workspace/pull/206/commits/a260201b556abbba352be2c9f73333700bd58e0b

[the remark this answers is not in the document]
Comment (aslakknutsen) [reply pkg/template/template_test.go]: Good point!

File: pkg/template/template_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply pkg/template/template_test.go]: originalDeployment and modifiedDeployment for the code and testDeployment for the test.

File: pkg/k8s/deployment.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply pkg/k8s/deployment.go]: ah, ok. that's great :)

**Pull request merged**